MIDAS/src/finite-state-machines/fsm.cpp

```diff
@@ -0,0 +1,369 @@
+#include <cmath>
+
+#include "fsm.h"
+#include "thresholds.h"
+#include "systems.h"
+
+
+// helper functions
+
+// most of these helper functions look to average the acceleration values inside the buffer
+
+float FSM::getAcceleration(std::array<HighGData, 8> & hg)
+{ 
+    float avg = 0;
+
+    for (size_t i = 0; i < hg.size(); i++)
+    {
+        avg += (hg.at(i).gz / (float)hg.size());
+    }
+
+    return avg;
+}
+
+float FSM::getAltitude(std::array<Barometer, 8> & bar)
+{
+    float avg = 0;
+
+    for (size_t i = 0; i < bar.size(); i++)
+    {
+        avg += (bar.at(i).altitude / (float)bar.size());
+    }
+
+    return avg;
+}
+
+double FSM::getJerk(std::array<HighGData, 8> & hg)
+{
+
+    double avg1 = 0;
+    for (size_t i = 0; i < hg.size() / 2; i++)
+    {
+        avg1 += (hg.at(i).gz / hg.size());
+    }
+
+    double avg2 = 0;
+    for (size_t i = hg.size() / 2; i < hg.size(); i++)
+    {
+        avg2 += (hg.at(i).gz / hg.size());
+    }
+
+    double sec = pdTICKS_TO_MS(hg.size()) * 1000;
+
+    return (avg2 - avg1) / (sec / 2);
+}
+
+double FSM::getVerticalSpeed(std::array<Barometer, 8> & bar)
+{
+    double avg1 = 0;
+    for (size_t i = 0; i < bar.size() / 2; i++)
+    {
+        avg1 += (bar.at(i).altitude / bar.size());
+    }
+
+    double avg2 = 0;
+    for (size_t i = bar.size() / 2; i < bar.size(); i++)
+    {
+        avg2 += (bar.at(i).altitude / bar.size());
+    }
+
+    double sec = pdTICKS_TO_MS((unsigned int)bar.size()) * 1000;
```

Comment:
For every place you calculate `sec` in this file, it won't quite work. We can't assume that one data point will be put in every tick (so using bar.size() isn't an accurate measure of time). Also, to convert milliseconds to seconds, you'd have to divide by 1000, not multiply.